fe/fe-core/src/main/java/com/starrocks/load/routineload/PulsarRoutineLoadJob.java

```diff
@@ -0,0 +1,567 @@
+            setPulsarPatitionInitialPositions(stmt.getPulsarPartitionInitialPositions());
+        }
+
+        if (!stmt.getCustomPulsarProperties().isEmpty()) {
+            setCustomPulsarProperties(stmt.getCustomPulsarProperties());
+        }
+    }
+
+    // this is an unprotected method which is called in the initialization function
+    private void setCustomPulsarPartitions(List<String> pulsarPartitions) {
+        this.customPulsarPartitions = pulsarPartitions;
+    }
+
+    private void setPulsarPatitionInitialPositions(List<Pair<String, Long>> patitionToInitialPositions) {
+        for (Pair<String, Long> entry : patitionToInitialPositions) {
+            ((PulsarProgress) progress).addPartitionToInitialPosition(entry);
+        }
+    }
+
+    private void setCustomPulsarProperties(Map<String, String> pulsarProperties) {
+        this.customProperties = pulsarProperties;
+    }
+
+    @Override
+    protected String dataSourcePropertiesJsonToString() {
+        Map<String, String> dataSourceProperties = Maps.newHashMap();
+        dataSourceProperties.put("serviceUrl", serviceUrl);
+        dataSourceProperties.put("topic", topic);
+        dataSourceProperties.put("subscription", subscription);
+        List<String> sortedPartitions = Lists.newArrayList(currentPulsarPartitions);
+        Collections.sort(sortedPartitions);
+        dataSourceProperties.put("currentPulsarPartitions", Joiner.on(",").join(sortedPartitions));
+        Gson gson = new GsonBuilder().disableHtmlEscaping().create();
+        return gson.toJson(dataSourceProperties);
+    }
+
+    @Override
+    protected String customPropertiesJsonToString() {
+        Gson gson = new GsonBuilder().disableHtmlEscaping().create();
+        return gson.toJson(customProperties);
+    }
+
+    @Override
+    public void write(DataOutput out) throws IOException {
+        super.write(out);
+        Text.writeString(out, serviceUrl);
+        Text.writeString(out, topic);
+        Text.writeString(out, subscription);
+
+        out.writeInt(customPulsarPartitions.size());
+        for (String partition : customPulsarPartitions) {
+            Text.writeString(out, partition);
+        }
+
+        out.writeInt(customProperties.size());
+        for (Map.Entry<String, String> property : customProperties.entrySet()) {
+            Text.writeString(out, "property." + property.getKey());
+            Text.writeString(out, property.getValue());
+        }
+    }
+
+    public void readFields(DataInput in) throws IOException {
+        super.readFields(in);
+        serviceUrl = Text.readString(in);
+        topic = Text.readString(in);
+        subscription = Text.readString(in);
+        int size = in.readInt();
+        for (int i = 0; i < size; i++) {
+            customPulsarPartitions.add(Text.readString(in));
+        }
+
+        if (GlobalStateMgr.getCurrentStateJournalVersion() >= FeMetaVersion.VERSION_51) {
```

Comment:
New features do not need to be compatible

Reply:
OK.